### ISSUE TITLE
[BUG] Visual Testing: Snapshots aren't capturing images

### ISSUE THREAD

**Issue by christopher-r-doucet** [microsoft/playwright#14388]
**Context:**
- Playwright Version: 1.2.2.
- Operating System: Ubuntu 20.0.4 Linux Subsystem for Windows 11 Home 10.0.22000
- Node.js version: v14.19.3
- Browser: Firefox, Mobile Safari
- Extra: 

**Code Snippet**

```
    test('Douglas 1', async ({ page }) => {
        await page.goto('https://yesibuybuyer1.carrot.com/', { waitUntil: 'networkidle' });
        expect(await page.screenshot({fullPage : true })).toMatchSnapshot();
      });
```

**Describe the bug**

I am trying to take golden snapshots of our production environment to compare to our staging environment for regression testing. The snapshots I'm taking are missing lazy loaded images for Firefox and Mobile Safari, causing invalid tests when run again. You can see down in the screenshots below where the images should load and where they are not.

I have checked in Browserstack the tests that are failing (Firefox, Mobile Safari) and the images are displayed correctly in Browserstack.

I've tried implementing waits, ```scrollToBottom`` etc. and its always the same issue with just those browsers.

**Add any other details about the problem here.**
![image](https://user-images.githubusercontent.com/8879741/170107840-858ea017-01d6-4d0d-9ae8-87879bd50c40.png)
![image](https://user-images.githubusercontent.com/8879741/170107887-2f9c628c-62d3-43ab-be27-ea22eb4e6878.png)
![image](https://user-images.githubusercontent.com/8879741/170108057-34e6d321-6353-4b34-865b-3fc9744addb1.png)
![image](https://user-images.githubusercontent.com/8879741/170108092-3496b745-2c70-4633-a58c-ad70bf989622.png)

**Comment by yury-s** [comment]
> I've tried implementing waits, ```scrollToBottom`` etc. and its always the same issue with just those browsers.

The image has `loading="lazy"` attribute so the browser may delay its load. You can add `await page.locator('img.wp-image-642').scrollIntoViewIfNeeded();` to make the image appear in the viewport and start loading but then you also need to wait for it to finish loading, so something like this should do:

```js
  const img = page.locator('img.wp-image-642');
  // kick off loading
  await img.scrollIntoViewIfNeeded();
  // wait for the image to finish loading
  await img.evaluate(image => image.complete || new Promise(f => image.onload = f));
```

We can probably add something like `expect(img).toBeLoaded();` but then you'd have to wait for each lazy image on the page manually.

**Comment by christopher-r-doucet** [comment]
> @christopher-r-doucet how many images do you have on a page?

@aslushnikov on a page like https://www.myhousefriend.com/ that I am trying to test there are 14. 
![image](https://user-images.githubusercontent.com/8879741/171452254-cbcdcb03-508e-4125-99fd-5b469ad865c4.png)
![image](https://user-images.githubusercontent.com/8879741/171453422-1ada7f37-5b1b-427c-a907-4d85ff570537.png)

**Comment by christopher-r-doucet** [comment]
@aslushnikov 

`test('Juniper 3', async ({ page }) => {
        await page.goto('https://alexmartinezseller2.carrot.com/');
        const img = page.locator('img.wp-image-1558');
        //const img2 = page.locator('img.wp-image-1564');
        //const img3 = page.locator('img.wp-image-1565');
        const img4 = page.locator('img.wp-image-1704');
        const img5 = page.locator('img.wp-image-1563');
        //const img6 = page.locator('img.wp-image-1564');
        //const img7 = page.locator('img.wp-image-1565');
        const img8 = page.locator('img.wp-image-1566');
        const img9 = page.locator('img.wp-image-1567');
        const img10 = page.locator('img.wp-image-1568');
        const img11 = page.locator('img.wp-image-1652');
        const img12 = page.locator('img.wp-image-1653');
        const img13 = page.locator('img.wp-image-1660');
        await img.scrollIntoViewIfNeeded();
        //await img2.scrollIntoViewIfNeeded();
        //await img3.scrollIntoViewIfNeeded();
        await img4.scrollIntoViewIfNeeded();
        await img5.scrollIntoViewIfNeeded();
        //await img6.scrollIntoViewIfNeeded();
        //await img7.scrollIntoViewIfNeeded();
        await img8.scrollIntoViewIfNeeded();
        await img9.scrollIntoViewIfNeeded();
        await img10.scrollIntoViewIfNeeded();
        await img11.scrollIntoViewIfNeeded();
        await img12.scrollIntoViewIfNeeded();
        await img13.scrollIntoViewIfNeeded();
        await img.evaluate(image => image.complete || new Promise(f => image.onload = f));
        //await img2.evaluate(image => image.complete || new Promise(f => image.onload = f));
        //await img3.evaluate(image => image.complete || new Promise(f => image.onload = f));
        await img4.evaluate(image => image.complete || new Promise(f => image.onload = f));
        await img5.evaluate(image => image.complete || new Promise(f => image.onload = f));
        //await img6.evaluate(image => image.complete || new Promise(f => image.onload = f));
        //await img7.evaluate(image => image.complete || new Promise(f => image.onload = f));
        await img8.evaluate(image => image.complete || new Promise(f => image.onload = f));
        await img9.evaluate(image => image.complete || new Promise(f => image.onload = f));
        await img10.evaluate(image => image.complete || new Promise(f => image.onload = f));
        await img11.evaluate(image => image.complete || new Promise(f => image.onload = f));
        await img12.evaluate(image => image.complete || new Promise(f => image.onload = f));
        await img13.evaluate(image => image.complete || new Promise(f => image.onload = f));
        expect(await page.screenshot({fullPage : true })).toMatchSnapshot();`

When I run this in my WSL Ubuntu environment it hangs, I tried it on my other machine which is just a standard Ubuntu OS sideboot with Windows and it seems to be running fine.

**Comment by aslushnikov** [comment]
> When I run this in my WSL Ubuntu environment it hangs, I tried it on my other machine which is just a standard Ubuntu OS sideboot with Windows and it seems to be running fine.

@christopher-r-doucet So this implies there's something wrong with your WSL. Is it WSL or WSL2? Note that things might be terribly slow under WSL, but should work fine in WSL2!

**Comment by pavelfeldman** [comment]
Maybe running those in parallel would help you speed this up:

```js
// Trigger loading of all images
await locators.evaluateAll(e => e.scrollInfoView());
// Set up listeners concurrently
const promises = locators.map(locator => locator.evaluate(image => image.complete || new Promise(f => image.onload = f)));
// Wait for all once
await Promise.all(promises);
```

**Comment by pavelfeldman** [comment]
Closing as per above, please feel free to open a new issue if this does not cover your use case.
